google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -144,135 +206,152 @@ public static MonitoredResource getResource(String projectId, String resourceTyp
-        value = System.getenv("K_SERVICE");
+      case RevisionName:
+        value = getter.getEnv("K_REVISION");
         break;
       case VersionId:
-        value = getAppEngineVersionId();
+        value = getter.getEnv("GAE_VERSION");
         break;
       case Zone:
-        value = MetadataConfig.getZone();
-        break;
-      default:
-        value = null;
+        value = getZone();
         break;
     }
+
     return value;
   }
 
-  /* Detect monitored Resource type using environment variables, else return global as default. */
-  private static Resource getAutoDetectedResourceType() {
-    if (System.getenv("K_SERVICE") != null
-        && System.getenv("K_REVISION") != null
-        && System.getenv("K_CONFIGURATION") != null
-        && System.getenv("KUBERNETES_SERVICE_HOST") == null) {
-      return Resource.CloudRun;
-    }
-    if (System.getenv("GAE_INSTANCE") != null) {
-      return Resource.GaeAppFlex;
-    }
-    if (System.getenv("KUBERNETES_SERVICE_HOST") != null) {
-      return Resource.K8sContainer;
-    }
-    if (ServiceOptions.getAppEngineAppId() != null) {
-      return Resource.GaeAppStandard;
-    }
-    if (MetadataConfig.getInstanceId() != null) {
-      return Resource.GceInstance;
+  /**
+   * Heuristic to discover the namespace name of the current environment. There is no determenistic
+   * way to discover the namespace name of the process. The name is read from the {@link
+   * K8S_POD_NAMESPACE_PATH} when available or read from a user defined environment variable
+   * "NAMESPACE_NAME"
+   *
+   * @return Namespace name or empty string if the name could not be discovered
+   */
+  private static String getK8sNamespace() {
+    String value = "";
+    try {
+      value =
+          new String(Files.readAllBytes(Paths.get(K8S_POD_NAMESPACE_PATH)), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      // if SA token is not shared the info about namespace is unavailable
+      // allow users to define the namespace name explicitly
+      value = getter.getEnv("NAMESPACE_NAME");
+      if (value == null) {
+        value = "";
+      }
     }
-    // default Resource type
-    return Resource.Global;
-  }
-
-  private static String getAppEngineModuleId() {
-    return System.getenv("GAE_SERVICE");
+    return value;
   }
 
-  private static String getAppEngineVersionId() {
-    return System.getenv("GAE_VERSION");
+  /**
+   * Distinguish between Standard and Flexible GAE environments. There is no indicator of the
+   * environment. The path to the startup-script in the metadata attribute was selected as one of
+   * the new values that explitly mentioning "flex" and cannot be altered by user (e.g. environment
+   * variable). The method assumes that the resource type is already identified as {@link
+   * Resource.AppEngine}.
+   *
+   * @return "flex" {@link String} for the Flexible environment and "standard" for the Standard.
+   */
+  private static String getAppEngineEnvironment() {
```

Comment:
is this tested? Maybe worth reaching out to serverless folks (DPE/SWE) to see if this is the best heuristic we can get

Reply:
There are two unit tests that validate discovery of the GAE resource type - one for flexible and another for standard environments.
The info is taken after careful inspection of what environments provide including environment variables, metadata attributes and file system. I choose smth that seemed more or less deterministic. The "legacy" code did not work well. If you like, I will demo you on Monday.

Reply:
Yes, would love to see a demo. Also worth confirming with @grayside - AppEngine folks work on Titanium release - I want to make sure this logic will continue to work in light of that work